cpp/include/raft/neighbors/ivf_pq_helpers.cuh

```diff
@@ -407,5 +575,160 @@ void erase_list(raft::resources const& res, index<IdxT>* index, uint32_t label)
+ * NB: This is to be used only when the rotation matrix is not already computed through
+ * raft::neighbors::ivf_pq::build.
+ *
+ * Usage example:
+ * @code{.cpp}
+ *   // compute the rotation matrix with random_rotation
+ *   raft::neighbors::ivf_pq::helpers::make_rotation_matrix(res, &index, true);
+ * @endcode
+ *
+ * @tparam IdxT
+ * @param[in] res
+ * @param[inout] index
+ * @param[in] force_random_rotation whether to apply a random rotation matrix on the input data. See
+ * raft::neighbors::ivf_pq::index_params for more details.
+ */
+template <typename IdxT>
+void make_rotation_matrix(raft::resources const& res,
+                          index<IdxT>* index,
+                          bool force_random_rotation)
+{
+  raft::neighbors::ivf_pq::detail::make_rotation_matrix(res,
+                                                        force_random_rotation,
+                                                        index->rot_dim(),
+                                                        index->dim(),
+                                                        index->rotation_matrix().data_handle());
+}
+
+/**
+ * @brief Public helper API for externally modifying the index's IVF centroids.
+ * NB: The index must be reset before this. Use raft::neighbors::ivf_pq::extend to construct IVF
+ lists according to new centroids.
+ *
+ * Usage example:
+ * @code{.cpp}
+ *   // allocate the buffer for the input centers
+ *   auto cluster_centers = raft::make_device_matrix<float, uint32_t>(res, index.n_lists(),
+ index.dim());
+ *   ... prepare ivf centroids in cluster_centers ...
+ *   // reset the index
+ *   reset_index(res, index);
+ *   // recompute the state of the index
+ *   raft::neighbors::ivf_pq::helpers::recompute_internal_state(res, index);
+ *   // Write the IVF centroids
+ *   raft::neighbors::ivf_pq::helpers::set_centers(
+                    res,
+                    &index,
+                    cluster_centers);
+ * @endcode
+ *
+ * @tparam IdxT
+ * @param[in] res
+ * @param[inout] index
+ * @param[in] cluster_centers the new cluster centers
+ */
+template <typename IdxT>
+void set_centers(raft::resources const& res,
+                 index<IdxT>* index,
+                 device_matrix_view<const float, uint32_t> cluster_centers)
+{
+  RAFT_EXPECTS(cluster_centers.extent(0) == index->n_lists(),
+               "Number of rows in cluster centers and IVF centers are different");
+  RAFT_EXPECTS(cluster_centers.extent(1) == index->dim(),
+               "Number of columns in cluster centers and index dim are different");
+  RAFT_EXPECTS(index->size() == 0, "Index must be empty");
+  ivf_pq::detail::set_centers(res, index, cluster_centers.data_handle());
+}
+
+/**
+ * @brief Helper to fetch size of a particular IVF list in bytes using the list extents.
+ *
+ * Usage example:
+ * @code{.cpp}
+ *   // Fetch the size of the fourth list
+ *   raft::neighbors::ivf_pq::helpers::get_list_size_in_bytes(index, 3);
```

Comment:
Please expand the usage examples here. We are doing this for all new public API functions for users to be able to copy/paste/modify. Copying the function prototype is not a usage example. Please add things like building the index in there. How is this index supposed to be uilt? What assumptions does it have? It's okay to see use "see also" tags in here as well, but we need to be making this useful for users who aren't familiar w/ our code or CUDA in general.

Reply:
Added a usage example with the current state of the index with populated IVF lists.